### ISSUE TITLE
New ggplot2.discrete.fill option doesn't work with ordered factor columns

### ISSUE THREAD

**Issue by rlh1994** [tidyverse/ggplot2#4148]
I have found what I think is an issue with the new ggplot2.discrete.fill options, although it may be my understanding of what exactly "discrete" means and this may be by design. When using the option it works correctly for character columns, or as in the example factor(char_column)'s, but does not appear to work for already existing factor columns as this uses the default colours instead.

As I said this may be by design in that factors are treated different in plotting and I haven't realised, but this isn't clear in documentation anywhere I looked. I am using ggplot2 version 3.3.2, and R version 3.6.1. 

``` r
library(ggplot2)
okabe <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

cty_by_var <- function(var) {
  ggplot(mpg, aes(cty, colour = {{var}}, fill = {{var}})) +
    geom_density()
}

price_by_var <- function(var) {
  ggplot(diamonds, aes(price, colour = {{var}}, fill = {{var}})) +
    geom_density()
}

price_by_var2 <- function(var) {
  ggplot(diamonds, aes(price, colour = as.character({{var}}), fill = as.character({{var}}))) +
    geom_density()
}

# Works correctly with okabe colours 
withr::with_options(
  list(ggplot2.discrete.fill = okabe),
  print(cty_by_var(class))
)
```

![](https://i.imgur.com/lLHXxJB.png)

``` r

# Works incorrectly, using the default colours
withr::with_options(
  list(ggplot2.discrete.fill = okabe),
  print(price_by_var(cut))
)
```

![](https://i.imgur.com/ML8IKNY.png)

``` r

# Works correctly with okabe colours 
withr::with_options(
  list(ggplot2.discrete.fill = okabe),
  print(price_by_var2(cut))
)
```

![](https://i.imgur.com/AfICFVi.png)

<sup>Created on 2020-07-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by yutannihilation** [comment]
Thanks, confirmed. Curious... (I replaced your reprex with the rendered version).

**Comment by yutannihilation** [comment]
Just remembered. The default colour/fill scales for ordered factors are different from the other. But, I think the option should work for ordered factors.

``` r
library(ggplot2)
okabe <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

withr::with_options(
  list(ggplot2.discrete.fill = okabe),
  print(
    ggplot(mpg, aes(cty, fill = factor(class))) +
      geom_density()
  )
)
```

![](https://i.imgur.com/rYxNtZM.png)

``` r

withr::with_options(
  list(ggplot2.discrete.fill = okabe),
  print(
    ggplot(mpg, aes(cty, fill = factor(class, ordered = TRUE))) +
      geom_density()
  )
)
```

![](https://i.imgur.com/pbmvlkH.png)

<sup>Created on 2020-07-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by yutannihilation** [comment]
A bit more context.

First two of this type of options were `ggplot2.continuous.{colour,fill}`, which were introduced in #2178. These were dedicated options to choose viridis palette as default or not.

`scale_{colour,fill}_ordinal()`, which are the default scale for ordered factors, were introduced in #2179. The intent of this is to distinguish ordered factors from unordered ones.

https://github.com/tidyverse/ggplot2/blob/b4bc293fc0ed38d2002b0833c91a90b7b0e27e63/R/zxx.r#L6
https://github.com/tidyverse/ggplot2/blob/b4bc293fc0ed38d2002b0833c91a90b7b0e27e63/R/zxx.r#L65

In short, at first, the system considered only two choices; viridis or not. Now that we have much more flexibility, we probably need more considerations. My take is:

* ggplot2 should have `ggplot2.ordinal.{colour,fill}` for consistency.
* `ggplot2.discrete.{fill,colour}` should affect (with a warning?) the fill/colour of ordinal variables when no `ggplot2.ordinal.{colour,fill}` is set.

**Comment by rlh1994** [comment]
I agree with your suggestions, having one option for all possible scale types feels correct. Same with discrete being the default when no ordinal is specified as I suspect many users (me included) aren't even aware of scale_*_ordinal.

**Comment by thomasp85** [comment]
I'm not sure I agree that the discrete colour setting should be used for ordinal data. What is your rationale for that? ordinal data is not very common so my guess is that when used, it is used for a reason and the scale should be sequential

**Comment by yutannihilation** [comment]
Since ordinal scale is a discrete scale, I thought it's natural that ordinal scale is subject to discrete scale's default. But, this is convincing...

> the scale should be sequential

**Comment by dgkf** [comment]
Thanks for tagging #4253, @yutannihilation. One added motivation in that issue is to allow for customization via `thematic` themes. Currently the `thematic` package doesn't touch ordinal scales, which causes the viridis defaults to stand out against some color pallets.

**Comment by cpsievert** [comment]
Since I feel partly responsible for this, I took a quick stab at this in #4277. @yutannihilation / @thomasp85  let me know if you think this is worth taking on.
